man/send.1
.\" $Header$
.\"
.\"  Man page for HUJI-NJE/FUNET-NJE on UNIX system.
.\"
.\"  Written by  Matti Aarnio <mea@finfiles.bitnet> <mea@nic.funet.fi>
.\"
.\"  Date: 27-Dec-1990, 18-Sep-1993, 10-Oct-1993
.\"
.ds ]W Funet-Nje 1 UNIX 3.0
.TH SEND 1
.SH NAME
send, ygone \- bitnet message send tools
.SH SYNOPSIS
.IP send 0.5i 0
.RB [ "\-s|\-u username" ]
.BI "\-c{ommand}" " host "
[ text of command ]
.IP send 0.5i 0
.RB [ "\-s|\-u username" ]
.RB [ "\-c\fR{\fBommand" }]
.I " @host "
[ text of command ]
.IP send 0.5i 0
.RB [ "\-s|\-u username" ]
.RB [ "\-m\fR{\fBessage" }]
.I " user@host "
[ text of message ]
.IP ygone 0.8i 0
.BR [ " \-d " ]
.SH DESCRIPTION
.IX "send"
.IX "bitnet"
.IX "funetnje"
.IX "sendfile"
.IX "submit"
.IX "netdata"
.IX "ygone"
.LP
.I send
sends NMR messages over RSCS network to target
.I host/user.
Only necessary argument is the first one telling if this is
.I command
to be sent to a hostname without giving
.RB ` @ '
in its name.
Commands go to (usual name) `RSCS' subsystem at each host,
Messages to specified users.
.LP
Examples:
.LP
.nf
.fp 4 CW
\f4send @bitnic cpq time
\f4send -c bitnic cpq time
\f4    \fR(Sends a "CPQ TIME" to BITNIC)
\f4send user@bitnic hello
\f4    \fR(Sends a "hello" to user@bitnic)
\f4send -u foo user@bitnic hello
\f4    \fR(Sends a "hello" to user@bitnic from user foo@<localhost>)
\f4send -s user@bitnic hello
\f4    \fR(Sends a "hello" to user@bitnic from local system: "@<localhost>")
.fi
.LP
When
.I send
is left without
.RI [ "text of message" ]
portion of arguments, it prompts that from user giving target address
as prompt string.  Hit empty line or EOF to quit.
.LP
If no 
.RI [ "text of message" ]
is given,
.I STDIN
is used as an input for 
.IR send ,
thus IO redirection can be used.
.LP
When
.I send
is left without target address, that too is asked from the user.
.LP
.I Ygone
controls temporary disposition of incoming NMR-messages into
a personal logfile.  In UNIX systems that file is
.I .ygone.messages
in user's home directory.
This association is broken when
.BR funetnje (8)
is restarted, as the list of service requesters is kept only in core..
.LP
With option
.B -d
user can disable
.IR ygone ,
and look at latest set of arrived messages (if any).
.LP
.SH ENVIRONMENT
With environment variable
.B EBCDICTBL
user can select alternate EBCDIC <\-> ASCII translation
tables.  For more info, see
.BR ebcdictbl (5).
.LP
.SH FILES
.I /etc/funetnje.cf
.SH SEE ALSO
.BR bitspool (5),
.BR bmail (8),
.BR ebcdictbl (5),
.BR funetnje (8),
.BR mailify (8),
.BR qrdr (8),
.BR receive (1),
.BR sendfile (1)
.SH BUGS
Doesn't always 
.LP
.SH DIAGNOSTICS
These programs show their usage summary, when there is something
wrong with input parameters.  Subtle usage bugs might not be detected!
.SH NOTES
.I Ygone
is subject to further development.
.SH AUTHORS
.LP
Yehavi Bourvine, Hebrew University of Yerusalem, Israel
.I <yehavi@hujivms.bitnet>.
.LP
Matti Aarnio, Finnish University and Research Network \- FUNET, Finland
.I <mea@nic.funet.fi> <mea@finfiles.bitnet>.
.LP
Gerald Hanusch, Johannes Kepler Universitaet/Universitaet Linz, Austria
<k000165@alijku11.bitnet>.
.LP
Original (VMS oriented package) software by Yehavi Bourvine,
UNIX features, plenty of modifications, file handling exits,
man-pages by Matti Aarnio.
Port of ygone to FUNET's version by Gerald Hanusch.


ChangeLog
Fri Mar 22 13:48:30 1996  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    Even more relaxed approach on parametrizing
	    the CMDMAILBOX -- now allows SOCK/FIFO/UDP
	    with fairly flexible (in regards of spaces)
	    syntax..

Mon Nov 13 17:48:52 1995  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    A bit more relaxed approach on parametrizing
	    CMDMAILBOX config parameter (FIFO/SOCKET case)

	* file_queue.c:
	    Do RSCS-accounting on locally sent files, that
	    wind up to local recipients

	* unix_brcst.c:
	    Re-enter the SysVR4 UTMP weirdos..  Now fixed ?

Fri Aug 25 09:02:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* mailify.c:
	    Handling error-messages (where "MAIL FROM:" address
	    is "<>", that is, empty) was done improperly.  Now
	    it does flag input as from "channel error" in that
	    case.

Mon Jul 24 14:29:59 1995  Matti Aarnio  <mea@nic.funet.fi>

	* bmail.c, sendfile.c:
	    Sometimes   mktemp()  fails when there are a LOT of
	    files being sent at the same time.  Changed now to
	    more classical  ".MaiL.%05d"-format with  getpid()
	    result as integer data..

Wed Jun 21 22:42:39 1995  Matti Aarnio  <mea@nic.funet.fi>

	* unix_brcst.c, nmr_unix.c:
	    Ok, more than __linux__ is needed to know which
	    machines have <utmp.h> containing extended flags.
	    Now we test  "DEAD_PROCESS" pre-processor define,
	    which existence (propably) will correlate with
	    the existence of utmp's field  ut_type.
	    (SysVr4 feature, I think.)

Mon Jun 19 15:17:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* send.c, sendfile.c, bmail.c:
	    Some additional tests to trap various forms of bad
	    addresses -- accepts only  user@node format with
	    no '.' in the address, nor either component longer
	    than 8 chars.

	* nmr.c:
	    Made sure not to overwrite allocated string spaces on
	    Msg-queue entry.  (Did cause a crash with it..)

Tue Jun 13 09:13:18 1995  Matti Aarnio  <mea@nic.funet.fi>

	* qrdr.c:
	    Collect directory browsing results to an array at first,
	    then qsort() it with the time-stamp, and finally display
	    them in order. (Latest last)
	    (This is something what Gerald Hanush asked for.)

Mon Jun 12 22:44:00 1995  Matti Aarnio  <mea@nic.funet.fi>

	* file_queue.c:
	    Apparently OSF/1 pipe() return value is the first
	    fd, not "0", when it succeeds.  Nice thing - and not..
	    Altered the test to see if it returns a value under
	    zero, or something positive.  When the return is
	    positive (zero, or larger), fork() and run in
	    child-mode.

	    If a child_pipe exists already ( >= 0 ), close
	    it and mark the new one for processing.

	* libreadcfg.c, read_config, libsendcmd.c, unix.c:
	    Rewrote things so that now the used CMDMAILBOX
	    transport method is chosen at runtime:

	    CMDMAILBOX  F  /path/to/.cmdfifo
	    CMDMAILBOX  f  /path/to/.cmdfifo
	    CMDMAILBOX  S  /path/to/.cmdsocket
	    CMDMAILBOX  U  127.0.0.2 175	# UDP port

	    The "f" is FIFO which is closed, and reopened
	    every time one command is pulled in.

Sun Jun 11 03:12:06 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesparser.c:
	    Noticed that the data parsing stopped at first ":",
	    which caused premature breaks; observe following entry
	    from BITEARN.NODES -file:
	    	:remark1.BASE(9501) 00/00/00 00:00:00

	* libreadcfg.c, read_config.c, libsendcmd.c, unix.c:
	   Amended "CMDMAILBOX"-data to contain port-number in
	   case UDP-format is used. (That is, have two tokens.)
	   Default value for the port-number is 175.

	* unix_tcp.c read_passive_tcp_connection():
	   Noticed that a case when a socket exists, and a new
	   connection is attempted is handled improperly. The
	   old socket-fd was FORGOTTEN!  Now it NAKs the connection
	   attempt with a complaint that there already is a connection
	   in there...

	   Things will progress from such a situation usually by
	   detecting a closure on the existing stream, but better
	   not to rush ahead of things.

	* bmail.c:
	   Accept "-u" only when calling the program from uids 0 or 1
	   (root and daemon in most common cases -- for BSD-sendmail
	    which wants  bmail executable with protection 2111 (user X,
	    group NJEGRP) )

Sat Jun 10 05:18:27 1995  Matti Aarnio  <mea@nic.funet.fi>

	* Makefile, consts.h, prototypes.h, recv_file.c, io.c,
	  send_file.c, unix.c, protocol.c, unix_tcp.c:

	    - Found one uninitialized pointer, which apparently
	      caused mystic data-structure corruptions (on timer
	      subsystem)

	    - Made mandatory (well, with disabling options) to
	      use NBSTREAM, USE_XMIT_QUEUE, and USE_SOCKOPT.
	      (in consts.h, see also  Makefile's  comments!)

	    - Use  F_WAIT_V_A_BIT  to flag if there is any reason
	      for NOT to send out data at the moment, and instead
	      quit from within  handle_ack()..
	      The F_WAIT_V_A_BIT is set when new data is sent to
	      outbound buffer (primary one, or XMIT QUEUE).
	      The bit is cleared when pending writer finishes
	      with success status.  When the pending-writer is
	      called from  poll_sockets(), immediately after it
	      the  dequeue_xmit_queue() is called, and if there
	      is something, it can cause F_WAIT_V_A_BIT to be
	      set again..)

	    - If the XMIT_QUEUE ever runs out of space, it abort()s!

	    - Tested the beast by sending 80 or so smallish files both
	      ways over the link.  Such setup caused previously link
	      corruption..  Actually the critical phase is when there
	      are flowing NJE headers/trailer, bare data sending works
	      just fine, but when a file ends, EOF block needs an ack
	      to be sent, and unless there the care is used, all
	      queues will overflow..

Thu Jun  8 15:48:14 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesfilter.c:
	    New option: "-netsoft" to list node software fields.

	* funetnje.texinfo:
	    Some additional chapters

	* unix_tcp.c:

send.c
/* SEND.C	V1.3-mea
 | Copyright (c) 1988,1989,1990 by
 | The Hebrew University of Jerusalem, Computation Center.
 |
 |   This software is distributed under a license from the Hebrew University
 | of Jerusalem. It may be copied only under the terms listed in the license
 | agreement. This copyright message should never be changed or removed.
 |   This software is gievn without any warranty, and the Hebrew University
 | of Jerusalem assumes no responsibility for any damage that might be caused
 | by use or misuse of this software.
 |
 | Send a command or message to the NJE emulator to send it over to the
 | correct line.
 | The call to this module is either by SEND/COMMAND or SEND/MESSAGE. If the
 | user gives the message's text, then we send it and exit. If not, we
 | inquire it interactively.
 | Since the command given is parsed by the shell as arguments, we have to
 | collect them back as one text line.
 | One line usage: send /command node-name command
 |            or:  send /message User@Node message...
 | Please note the space between SEND and the qualifier.
 |
 */

#include "prototypes.h"
#include "clientutils.h"

#define LINESIZE 512
#define MAXTEXTSIZE 103 /* Should be able to use at least 103 chars.. */
#define CMD_MSG 0
#define CMD_CMD 1

void
usage(explanation)
char *explanation;
{
	if (!explanation) {
	  fprintf(stderr,"Usage: send [-u fromuser|-s] [-c{ommand}] [@]node [command string]\n");
	  fprintf(stderr,"       send [-u fromuser|-s] [-m{essage}] user@node [message text]\n");
	} else {
	  fprintf(stderr,"SEND: %s\n",explanation);
	}
	exit(9);
}


char LOCAL_NAME   [10];
char BITNET_QUEUE [80];
char LOG_FILE     [80] = "-";	/* STDERR as the default */
int silent_wordwrap = 0;

int LogLevel = 1;
FILE *LogFd = NULL;

extern int send_nje();

int
main(cc, vv)
char	**vv;
int	cc;
{
	char	text[LINESIZE];
	char	address[32];
	char	from[32];
	char	*p;
	int	type = 0,	/* Message or command */
		mode = 0;	/* Interactive or not */
	struct stat stdstats;
	int	noprompt = 0;


	setbuf (stdout,NULL);	/* No buffering on stdout,
				   no need to flush... */
	fstat(fileno(stdin),&stdstats); /* This MUST succeed! */
	if (!S_ISCHR(stdstats.st_mode))
	  noprompt = 1;

	read_configuration();
	read_etable();
	*from = 0;

	/* Get the command line and command parameters.*/
	if (cc < 2) {	/* No /Mode or no address.. - tell user and exit */
	  usage(NULL);
	  exit(1);
	}

/* Get the switch */
	++vv;		/* Point to the qualifier */
	if (strcasecmp(*vv,"-u") == 0) {
	  ++vv;
	  cc -= 2;
	  if (*vv) {
	    strncpy(from,*vv,sizeof(from)-1);
	    from[sizeof(from)-1] = 0;
	    if ((p = strchr(from,'@')) != NULL) {
	      if ((p - from) > 8 || strlen(p+1) > 8) {
	    fromaddrerror:
		usage("Defined sender address (-u) components may not exceed 8 chars in size");
	      }
	    } else {
	      if (strlen(from) > 8)
		goto fromaddrerror;
	    }
	  } else {
	    usage("-u option requires one parameter!");
	  }
	  ++vv;
	}
	if (strcasecmp(*vv,"-s") == 0) {
	  ++vv;
	  --cc;
	  sprintf(from,"@%s",LOCAL_NAME);
	  if (getuid() >= LuserUidLevel) {
	    type = -1;
	    usage("-s option requires priviledges!");
	  }
	  type = CMD_MSG;
	} else if ((strcasecmp(*vv, "/message") == 0) ||
		   (strcasecmp(*vv, "-message") == 0) ||
		   (strcasecmp(*vv, "-m") == 0)) {
	  type = CMD_MSG;
	  --cc;
	  ++vv;
	} else if((strcasecmp(*vv, "/command") == 0) || 
		  (strcasecmp(*vv, "-command") == 0) ||
		  (strcasecmp(*vv, "-c") == 0)) {
	  type = CMD_CMD;
	  --cc;
	  ++vv;
	} else if (**vv == '/' || **vv == '-') {
	  usage("Valid qualifiers are `-c{ommand}' and `-m{essage}' only");
	} else
	  type = -1;	/* No qualified defined.. */

/* Get the address (if exists) */
	if (cc < 1) {	/* No parameters - prompt for address and enter interactive mode */
	  while (1) {
	    if (type == CMD_CMD)
	      printf("_Host_Address: ");
	    else if (type == CMD_MSG)
	      printf("_User@Host_Address: ");
	    else
	      printf("_Address: ");
	    if (fgets(address, sizeof address, stdin)== NULL) /* EOF */
	      exit(0);
	    address[sizeof(address)-1] = 0; /* Make sure it fills only so much.. */
	    if ((p = strchr(address, '\n')) != NULL) *p = '\0';
	    mode = 1;		/* Interactive mode */
	    if (*address == 0) continue;	/* Blank address.. */
	    if (*address == '@') {
	      strcpy(address,address+1);
	      if (strlen(address) > 8) {
	    addresserror:
		usage("target address may not have components longer than 8 chars");
	      }
	      type = CMD_CMD;
	      break;
	    } else if ((p = strchr(address,'@')) != NULL) {
	      if ((p - address) > 8 || strlen(p+1) > 8)
		goto addresserror;
	      type = CMD_MSG;
	      break;
	    } else if (type == -1) { /* Not defined! */
	      if (strlen(vv[-1]) > 8)
		goto addresserror;
	      type = CMD_MSG;
	      break;
	    }
	    break;
	  }
	} else {		/* Get the available parameters */

	  if (!*vv)
	    usage("Mandatory parameter (target address) is missing!");

	  strncpy(address, *vv, sizeof(address)); /* We have at least
						     the address there */
	  address[sizeof(address)-1] = 0;	  /* Make sure it fills
						     only so much.. */
	  ++vv;
	  --cc;

	  if (*address == '@') {
	    strcpy(address,address+1);
	    if (strlen(address) > 8)
	      goto addresserror;
	    type = CMD_CMD;
	  } else if ((p = strchr(address,'@')) != NULL) {
	    if ((p - address) > 8 || strlen(p+1) > 8)
	      goto addresserror;
	    type = CMD_MSG;
	  } else if (type == -1) {
	    if (strlen(vv[-1]) > 8)
	      goto addresserror;
	    sprintf(address,"%.8s@%.8s",vv[-1],LOCAL_NAME);
	    type = CMD_MSG;
	  }

	  if (cc <= 1) {	/* Nothing more than it		*/
	    mode = 1;		/* Interactive mode		*/
	  } else {		/* Reconstruct the parameters as the text */
	    char *tout = text;
	    int  charspace = sizeof(text)-2;
	    *text = '\0';
	    while (cc > 0 && *vv) {
	      char *s = *vv;
	      while (*s && charspace > 0) {
		*tout++ = *s++;
		--charspace;
	      }
	      if (vv[1] != NULL && charspace > 0)  /* More to come ? */
		*tout++ = ' ';
	      *tout = 0;
	      ++vv; --cc;
	    }
	  }
	}

	/* Create the sender's address */
	if (getuid() >= LuserUidLevel && *from ) {
	  fprintf(stderr,"SEND: -u option not allowed, insufficient priviledges.\nForcing uname to your login id.   Your uid = %d\n",(int)getuid());
	  *from = 0;
	}
	if (*from == 0)
	  mcuserid(from);

	if (strchr(address,'.') != NULL)
	  usage("target address may not contain '.' chars in it!");
	if (strchr(from,'.') != NULL)
	  usage("source address may not contain '.' chars in it!");

	if (mode == 0) {		/* Batch mode */
	  send_nje(type, from, address, text);	/* Send it to the daemon */
	  exit(0);
	}

/* We have to read it interactively. Loop until blank line */
	if (!noprompt)
	  fprintf(stderr,
		  "Hit your message/command. End with EOF (usually Ctrl-D)\n");
	while (!feof(stdin) && !ferror(stdin)) {	/* [mea] */
	  if (!noprompt)
	    printf("%s: ", address);
	  *text = 0;
	  if (fgets(text, sizeof(text)-1, stdin) == NULL)
	    break;		/* EOF */
	  text[sizeof(text)-1] = 0;
	  if ((p = (char*)strchr(text, '\n')) != NULL) *p = '\0';
	  send_nje(type, from, address, text);
	}
	exit(0);
}


int
send_nje(type, from, address, text)
int	type;
char	*from, *address, *text;
{
	char	line[140], *p;
	int	size;
	int	textlen, addrlen;

	/* Remove all controls */
	for (p = text; *p != '\0'; p++)
	  if (((unsigned char)*p) < ' ')
	    *p = ' ';

	if (type == CMD_CMD) {
	  /* Uppercase the message's text */
	  upperstr(text);
	  *line = CMD_SEND_COMMAND;
	}
	else
	  *line = CMD_SEND_MESSAGE;

	if (type == CMD_CMD)	/* Add the @ to address */
	  sprintf(&line[2], "%s @%s ", from, address);
	else			/* Address already has @ */
	  sprintf(&line[2], "%s %s ", from, address);

	addrlen = strlen(line+2);

	/* Upper-case the addresses.
	   If this is a command - uppercase it also */
	upperstr(line+2);
	if (type == CMD_CMD)
	  upperstr(text);

	textlen = strlen(text);

	line[1] = addrlen + 2;		/* Where the text begins */
	if (textlen <= MAXTEXTSIZE) {
	  strcpy(line+2+addrlen, text);

	  size = 3+addrlen+textlen;	/* 3 for the command code,
					   text start offset,
					   and string end NULL */

	  return send_cmd_msg(line, size, 0); /* Must be ONLINE! */

	} else {
	  /* Ok, so the text length is more than the allowed maximum
	     (103 chars), do wraps et.al. */
	  char *t = text;

	  /* XX: Complain about long input ?? */

	  while (textlen > 0) {
	    int partlength = textlen > MAXTEXTSIZE  ? MAXTEXTSIZE : textlen;
	    int rc;

	    strncpy(line+2+addrlen, t, partlength);
	    size = 3+addrlen+partlength;
	    line[size-1] = 0; /* Add terminating NULL */

	    rc = send_cmd_msg(line, size, 0); /* Must be ONLINE! */
	    if (rc) return rc;  /* Return immediately,
				   if sending was not ok.. */

	    t += partlength;
	    textlen -= partlength;
	  }
	  return 0;
	}
}
